### DISCUSSION TITLE
UDM Pro 3.x - nspawn  Installation - Questions

### DISCUSSION THREAD

**Discussion by Antergosgeek** [unifi-utilities/unifi-common#614] [Q&A] [answered]
Hi,

trying to follow this guide: https://github.com/unifi-utilities/unifios-utilities/tree/main/nspawn-container#step-2a-configure-the-container-to-use-an-isolated-macvlan-network

Says to create the VLAN in Unifi first. 
Any particular options? 
Isolate Network, I assume yes? How do I ensure other VLANS have access to it (or do they by default?)?
MutlicastDNS yes or no?
I already set VLAN ID to 5.
IP range 192.168.3.x I hope will be ok to follow the example in the guide when I use ```IPV4_GW``` and set the gateway interface's IP to 10.0.5.1/24, as well as ```IPV4_IP``` to the default  10.0.5.3? 10.0.5.x I do not use anywhere else on my Unifi.

I appreciate the help!

**Comment by m4n63** [comment]
By default routing will be in place, but you may need appropriate firewall rules.
The IP range / gw need to match, so if you use the examples in the guide the network in UDM will need to be configured as 10.0.5.0/24 with the UDM having .1 as its gateway ip.

**Comment by Antergosgeek** [comment]
Thank you! @m4n63 

Ok, that makes sense and I am ok with that as well so basically setup an isolated VLAN with 10.0.5.0/24 .

Like this: (screenshot from mobile)
![IMG_9290](https://github.com/unifi-utilities/unifios-utilities/assets/26154541/389ee333-9c5c-4228-9eef-7973e4cfd20a)


The part in brackets in the guide made me think I should use something different:
"Modify VLAN to an existing VLAN network that you want your container to be on. The default is VLAN 5. Make sure this VLAN network is created in Unifi first with a unique subnet and IP (do not use the same IP as you will use for IPV4_IP or IPV4_GW in this script)."

Correct like that?

**Reply by m4n63** [reply]
Yes a bit confusing that last part of the sentence, I agree.
Above should work I think.

**Reply by Antergosgeek** [reply]
Alright, thank you @m4n63. Will check back and report if it all worked. Thanks for your help!

**Comment by m4n63** [comment, marked as answer]
Sounds good!
and I just submitted a PR on this to update the documentation, I hope this makes it easier to understand when it is written in this way instead.
https://github.com/m4n63/unifios-utilities/blob/patch-3/nspawn-container/README.md
😎

**Reply by Antergosgeek** [reply]
Yes, I think that’s perfect!

**Comment by Antergosgeek** [comment]
Ok so my container is running and showing the right IP, I did however run into a few issues:

1. Neither of the curl commands worked for me:
```
curl -LO https://raw.githubusercontent.com/unifi-utilities/unifios-utilities/nspawn/nspawn-container/scripts/10-setup-network.sh
```

```
curl -LO https://raw.githubusercontent.com/unifi-utilities/unifios-utilities/nspawn/nspawn-container/scripts/0-setup-system.sh
```

both files just had "404 NOT FOUND" in them. So I tracked down the files myself and manually inserted the script contents in the files.

2. When I executed ```/data/on_boot.d/10-setup-network.sh``` I got:

![IMG_9292](https://github.com/unifi-utilities/unifios-utilities/assets/26154541/c10635b8-d12a-4c5f-9422-f2316a2898e0)

Hope that is not a problem?!

3. When I followed [Step 3: Configure Persistence Across Firmware Updates] (https://github.com/unifi-utilities/unifios-utilities/tree/main/nspawn-container#step-3-configure-persistence-across-firmware-updates) and ran ```apt download systemd-container libnss-mymachines debootstrap arch-test```
I got the following:

![IMG_9293](https://github.com/unifi-utilities/unifios-utilities/assets/26154541/5f0084f6-f7eb-4284-a8b0-a37ec1103ba6)

What should I do now? I assume this is important for fall-back and troubleshooting. Need some help with this one. Also how would I use the backup dpkg package files for systemd-container and dependencies once successfully downloaded in case I need them?

Thank you so much!

**Reply by m4n63** [reply]
regarding 1
broken URL's are starting to be common thing with this repository, send a PR or provide me the working links and I'll do one.

regarding 2
Not a propblem, just output from an IF clause in the script that you can ignore

regarding 3
just a warning as I read. I assume the file was still downloaded? 

Also, in my experience future firmware updates will sometimes break something no matter what you do to try and mitigate it so always be prepared for some manual fixes after each firmware update.

**Reply by Antergosgeek** [reply]
regarding 1:
I unfortunately don’t know how to do a PR, something I got to learn now that I am becoming more and more active on GitHub. I am a hobbyist…

but here are the links:
https://raw.githubusercontent.com/unifi-utilities/unifios-utilities/main/nspawn-container/scripts/0-setup-system.sh

https://raw.githubusercontent.com/unifi-utilities/unifios-utilities/main/nspawn-container/scripts/10-setup-network.sh

regarding 2:
Ok, thank you for explaining 

regarding 3:
Yes, file still downloaded.
How would I use these in case something does break?
When you say manually fixing stuff after each update, will it do things to the containers only or could it prevent the internet connection?
Any things to watch out for in particular?

I went ahead and installed pihole in the debian-custom nspawn container, everything works as expected! 

And lastly: 
@m4n63 "Thank you so much" for your help!

**Reply by Antergosgeek** [reply]
@m4n63 

Would you be able to guide me on this one:

regarding 3:
Yes, file still downloaded.
How would I use these in case something does break?
When you say manually fixing stuff after each update, will it do things to the containers only or could it prevent the internet connection?
Any things to watch out for in particular?

I kind of would like to know what to in case things stop working :-)

**Reply by m4n63** [reply]
In my experience of upgrading firmware, the data in the on_boot.d directlriy is always retained, but usually I have had to reinstall the udm-boot package as it is that package that triggers all these scripts to run.

**Reply by Antergosgeek** [reply]
Ok, that makes sense. Thank you so much!
